### PR TITLE
bpo-46841: Use inline caching for calls

### PR DESCRIPTION
**Pull request by brandtbucher**
Also:
- remove the "old" non-inline caching machinery
- fix some bugs in the collection of `PRECALL`/`CALL` specialization stats
- shrink the cache size requirements for `PRECALL`/`CALL` instructions

Next steps tracked at https://github.com/faster-cpython/ideas/issues/310.

https://bugs.python.org/issue46841

### CODE REVIEW

File: Include/internal/pycore_code.h [hunk 1]
```diff
@@ -112,73 +72,22 @@ typedef struct {
...
-static inline SpecializedCacheEntry *
-_GetSpecializedCacheEntryForInstruction(const _Py_CODEUNIT *first_instr, int nexti, int oparg)
-{
-    return _GetSpecializedCacheEntry(
-        first_instr,
-        offset_from_oparg_and_nexti(oparg, nexti)
-    );
-}
+/* Maximum size of code to quicken, in code units. */
+#define MAX_SIZE_TO_QUICKEN 10000
```
Comment (markshannon) [inline Include/internal/pycore_code.h:90]: Is this just to get the unpack sequence benchmark to work again, or something else?

File: Include/internal/pycore_code.h [hunk 2]
```diff
@@ -205,6 +114,13 @@ _Py_IncrementCountAndMaybeQuicken(PyCodeObject *code)
 
 extern Py_ssize_t _Py_QuickenedCount;
 
+// Borrowed references to common callables:
+struct callable_cache {
+    PyObject *isinstance;
+    PyObject *len;
+    PyObject *list_append;
+};
```
Comment (markshannon) [inline Include/internal/pycore_code.h:122]: I think the existence of `PyList_Type` as part of the API means that `list.append` must be per-process unique.
In other words, `list_append` could be static.

I'm happy to leave it as is for now, though. We should look to make the whole struct static, although the mutability of builtin functions makes that tricky for `isinstance` and `len`.

**Review by markshannon — APPROVED** [review]
Excellent. 
A couple comments, but nothing to block merging.

File: Include/internal/pycore_code.h (hunk 1 above)
Comment (brandtbucher) [reply Include/internal/pycore_code.h:90]: Nope, just for `unpack_sequence`.

File: Include/internal/pycore_code.h (hunk 2 above)
Comment (brandtbucher) [reply Include/internal/pycore_code.h:122]: I believe each interpreter has its own `builtins` module (check out `_PyBuiltin_Init`), so making this static could be tricky. As you said, though: probably worth looking into in the future.

**Pull request merged**